### PR TITLE
Matrix Cleanup 7

### PR DESCRIPTION
**Pull request by siefkenj**
This PR creates `MatrixCalculus` and `MatrixDecompositions` classes.  It is meant to be merged after PR #12708.

`MatrixDecompositions` houses the Cholesky, LDL, LU, and QR decompositions.  The performance of Cholesky, LDL, and QR on small matrices has been improved by a factor of 2, and for large matrices, the performance is similar.  As per bug #12389, `cholesky`, `LDLdecomposition`, `LUdecomposition`, `LUdecompositions_Simple`, `LUdecomposition_FF`, and `QRdecomposition` have been deprecated in favor of `cholesky_decomposition`, `LU_decomposition`, and `QR_decomposition`.  The other LU decompositions are accessible via keyword arguments of `method='compact'` for `LUdecomposition_Simple` and `method='ff'` for `LUdecomposition_FF`.

This patch also fixes the Cholesky bug #12500.

### CODE REVIEW

**Comment by jksuom** [comment]
matrices.py has grown inconveniently large. Would it be possible to form new modules out of some of the new matrix classes?

**Comment by siefkenj** [comment]
@jksuom do you mean moving things to separate classes?

**Comment by jksuom** [comment]
Yes, only suitable names should be invented. Otherwise I believe there should be no problems. That kind of move would seem appropriate now as the new grouping is being formed.

**Comment by siefkenj** [comment]
@jksuom I see a natural split for a `commonmatrix.py` which would have `MatrixRequired` and the 5 classes that only depend on it.  These classes should also work will all matrix-like objects, including ones over different rings (only __div__ uses division).  Would two files be the right split, or do you think it should be split over more files?

**Comment by jksuom** [comment]
It seems to me that two files would suffice at this point. What would remain if `MatrixRequired` and its descendants were moved over?

**Comment by siefkenj** [comment]
Only things that inherit strictly from `MatrixRequired` would move.  So `MatrixDeterminant`, `MatrixReductions`, etc. would remain.

I will wait for https://github.com/sympy/sympy/pull/12686 to be merged to do the split.  It is ready.

**Comment by jksuom** [comment]
I had been thinking that the sophisticated methods like those in `MatrixDeterminant` and `MatrixRedutions` would eventually be removed from matrices.py while more elementary methods would remain. That would probably mean that the split could not be based strictly on the class structure. Would that be a problem?

**Comment by siefkenj** [comment]
Do you think `MatrixBase` should still inherit everything, or should `MatrixBase` be kept to the simple stuff and Dense/Sparse matrices would inherit the advanced functionality directly?

**Comment by jksuom** [comment]
I am probably not sufficiently familiar with the code to give a good answer, but a quick glance seems to indicate that some of the advanced methods do not depend on the type of the input matrices. So it would appear that both dense and sparse matrices could inherit them via `MatrixBase`.

**Comment by siefkenj** [comment]
Looking at the diff, this seems really hard to review...any suggestions?

**Comment by siefkenj** [comment]
I've created a PR that is just splitting the files here: https://github.com/sympy/sympy/pull/12725  We'll try this in pieces.

**Comment by siefkenj** [comment]
Okay.  This should be much easier to review now.

File: sympy/matrices/sparse.py [hunk 1]
```diff
@@ -342,6 +342,42 @@ def _eval_is_symmetric(self, simpfunc):
         diff = (self - self.T).applyfunc(simpfunc)
         return len(diff.values()) == 0
 
+    def _eval_LDL_decomposition(self):
+        """Algorithm for numeric LDL factization, exploiting sparse structure.
```
Comment (jksuom) [inline sympy/matrices/sparse.py]: factorization

**Comment by jksuom** [comment]
This looks good to me. There are no essential changes.

**Comment by asmeurer** [comment]
Looks like some doctest failures were not caught by Travis

```
_____________________________________________________________ sympy.matrices.dense.jordan_block ______________________________________________________________
File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/matrices/dense.py", line 1297, in sympy.matrices.dense.jordan_block
Failed example:
    jordan_cell(x, 4)
Exception raised:
    Traceback (most recent call last):
      File "/Users/aaronmeurer/anaconda3/lib/python3.5/doctest.py", line 1321, in __run
        compileflags, 1), test.globs)
      File "<doctest sympy.matrices.dense.jordan_block[2]>", line 1, in <module>
        jordan_cell(x, 4)
      File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/core/decorators.py", line 38, in wrapper
        _warn_deprecation(wrapped, 3)
      File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/core/decorators.py", line 23, in _warn_deprecation
        SymPyDeprecationWarning(**decorator_kwargs).warn(stacklevel=stacklevel)
      File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/utilities/exceptions.py", line 154, in warn
        warnings.warn(see_above, SymPyDeprecationWarning, stacklevel=stacklevel)
    sympy.utilities.exceptions.SymPyDeprecationWarning:

    jordan_cell has been deprecated since SymPy 1.1. Use jordan_block
    instead.

______________________________________________________________________________________________________________________________________________________________
____________________________________________ sympy.matrices.matrices.MatrixDecompositions.cholesky_decomposition _____________________________________________
File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/matrices/matrices.py", line 1953, in sympy.matrices.matrices.MatrixDecompositions.cholesky_decomposition
Failed example:
    A.cholesky()
Exception raised:
    Traceback (most recent call last):
      File "/Users/aaronmeurer/anaconda3/lib/python3.5/doctest.py", line 1321, in __run
        compileflags, 1), test.globs)
      File "<doctest sympy.matrices.matrices.MatrixDecompositions.cholesky_decomposition[2]>", line 1, in <module>
        A.cholesky()
      File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/core/decorators.py", line 38, in wrapper
        _warn_deprecation(wrapped, 3)
      File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/core/decorators.py", line 23, in _warn_deprecation
        SymPyDeprecationWarning(**decorator_kwargs).warn(stacklevel=stacklevel)
      File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/utilities/exceptions.py", line 154, in warn
        warnings.warn(see_above, SymPyDeprecationWarning, stacklevel=stacklevel)
    sympy.utilities.exceptions.SymPyDeprecationWarning:

    cholesky has been deprecated since SymPy 1.1. Use
    cholesky_decomposition instead. See
    https://github.com/sympy/sympy/issues/12389 for more info.

**********************************************************************
File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/matrices/matrices.py", line 1958, in sympy.matrices.matrices.MatrixDecompositions.cholesky_decomposition
Failed example:
    A.cholesky() * A.cholesky().T
Exception raised:
    Traceback (most recent call last):
      File "/Users/aaronmeurer/anaconda3/lib/python3.5/doctest.py", line 1321, in __run
        compileflags, 1), test.globs)
      File "<doctest sympy.matrices.matrices.MatrixDecompositions.cholesky_decomposition[3]>", line 1, in <module>
        A.cholesky() * A.cholesky().T
      File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/core/decorators.py", line 38, in wrapper
        _warn_deprecation(wrapped, 3)
      File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/core/decorators.py", line 23, in _warn_deprecation
        SymPyDeprecationWarning(**decorator_kwargs).warn(stacklevel=stacklevel)
      File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/utilities/exceptions.py", line 154, in warn
        warnings.warn(see_above, SymPyDeprecationWarning, stacklevel=stacklevel)
    sympy.utilities.exceptions.SymPyDeprecationWarning:

    cholesky has been deprecated since SymPy 1.1. Use
    cholesky_decomposition instead. See
    https://github.com/sympy/sympy/issues/12389 for more info.
```

**Comment by asmeurer** [comment]
I am reverting this in the release branch at https://github.com/sympy/sympy/pull/12854. The deprecations need to be discussed further. If they are done, they need to be done according to our [deprecating policy](https://github.com/sympy/sympy/wiki/Deprecating-policy). Since @siefkenj is not free this week, I have reverted the whole PR (there weren't separate commits here for just the deprecations).

**Comment by siefkenj** [comment]
@asmeurer Are you referring to the fact that #12685 doesn't have discussion?

Depending on how cautious we are being about introducing API, the easiest fix for this would be to comment out the deprecation warnings for the methods we don't feel comfortable deprecating.  I can fix doc tests tomorrow.  If this is all too sudden, then I guess reverting is fine.  Though I hope this work doesn't get lost.  In general the matrix cleanup work fixes a ton of bugs that I didn't file (I didn't file most of the easy bugs like "... doesn't through an error for invalid matrix dimensions ..." or somesuch because I was worried about people trying to fix them prematurely).

**Comment by asmeurer** [comment]
I mean the fact that there were no deprecation issues opened explaining why the deprecations were made and how to fix the code. I also think we should be more conservative in deprecating things. 
I agree with @moorepants that too many deprecations can be a turnoff to users, and I don't really like deprecating function names just for no reason other than that they badly chosen for whatever reason. 

I didn't have time to work through this, and since you are busy, I just removed all matrix deprecations in https://github.com/sympy/sympy/pull/12854. We can go through them after the release and see what makes sense. And yes, sadly I had to revert this whole PR there too, since it was not easy to separate out just the deprecations, so any unrelated fixes will need to be made again (sorry about that. In the future, try to keep things as separate git commits). 

The way I did it, both old and new names exist, so deprecation will still be possible (in fact, I feel a lot better about deprecating something if the thing replacing it already has existed for a while). 

Sorry again for rushing this, but I need to get the release candidate out **now**.

**Pull request merged**